lib/galaxy/tools/deps/mulled/mulled_list.py

```diff
@@ -0,0 +1,168 @@
+    return [n for n in quay_list if n not in singularity_list and n not in blacklist]
+
+
+def get_missing_envs(quay_list, conda_list, blacklist_file=None):
+    r"""
+    >>> from os import remove
+    >>> with open('/tmp/blacklist.txt', 'w') as f:
+    ...     f.write('l\n\ng\nn\nr')
+    8
+    >>> get_missing_envs(quay_list=['1', '2', '3', 'h--1', 'g--2', 'r'], conda_list=['3', '4', '5'], blacklist_file='/tmp/blacklist.txt')
+    ['1', '2', 'h--1']
+    """
+    blacklist = []
+    if blacklist_file:
+        blacklist = open(blacklist_file).read().split('\n')
+
+    return [n for n in quay_list if n.split('--')[0] not in conda_list and n.split('--')[0] not in blacklist]
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description='Returns list of Docker containers in the quay.io biocontainers repository.')
+    parser.add_argument('--source', '-s',
+                        help="Docker, Singularity or Conda.")
+    parser.add_argument('--not-singularity', dest='not_singularity', action="store_true",
+                        help="Exclude Docker containers from which Singularity containers have already been built.")
+    parser.add_argument('--not-conda', dest='not_conda', action="store_true",
+                        help="Exclude Docker containers from which Conda environments have already been extracted.")
+    parser.add_argument('--conda-filepath', dest='conda_filepath', default=None,
+                        help="If searching for conda environments or employing the --not-conda option, a filepath where the environments are located.")
+    parser.add_argument('-b', '--blacklist', dest='blacklist', default=None,
+                        help="Provide a 'blacklist file' containing containers which should be excluded from the list.")
+    parser.add_argument('-f', '--file', dest='output', default=None,
+                        help="File to write list to. If not given output will be returned on the command line.")
+
+    args = parser.parse_args()
+
+    if args.source == 'docker':
+        containers = get_quay_containers()
+        if args.not_singularity:
+            containers = get_missing_containers(
+                containers, get_singularity_containers(), args.blacklist)
+        if args.not_conda:
+            containers = get_missing_envs(containers, get_conda_envs(
+                args.conda_filepath), args.blacklist)
+    elif args.source == 'singularity':
+        containers = get_singularity_containers()
+    elif args.source == 'conda':
+        containers = get_conda_envs(args.conda_filepath)
+    else:
+        print("The 'source' argument was not understood.")
+        return
+
+    if args.output:
+        with open(args.output, 'a') as f:
+            for container in containers:
+                f.write('%s\n' % container)
+    else:
+        print(containers)
+
+
+if __name__ == '__main__':
+    # main()
+
+    import doctest
+    doctest.testmod()
```

Comment:
Again, those doctests should probably live as unit tests in test/integration, or if you want, test/mulled